### PR TITLE
Add support for 2024-09-01 version of AKS API

### LINKED ISSUE #4164: Feature: Support for Microsoft.ContainerService/managedClusters/maintenanceConfigurations
The lack of this feature prevents the automatic upgrade configuration of AKS clusters.
Thank you!

### LINKED ISSUE #4327: Feature: Support newest AKS stable API
Current latest supported AKS API version is 2023-10-01
There have been a few other stable versions since then.  Request to update to latest 2024-05-01.

### PR DESCRIPTION
**Pull request by matthchr**
## What this PR does

- Fix issue in `RepairSkippingProperties` which could cause an issue when a type is reused among multiple resources, such as in a containerservice ManagedCluster and AgentPool.
- Update the rest-api-specs submodule, to get access to the hot off the presses 2024-09-01 version of the containerservice API.
- Add new `StripLargeDescriptions` pipeline stage, which can be used to strip descriptions off CRD types to reduce their size.
  - We don't want to do this in perpetuity, see #4418 which tracks removing some old versions of ManagedCluster so we can drop this in a future release.
 - Add support for 2024-09-01 version of AKS API.
 - Add support for MaintenanceConfiguration.

Fixes #4164.
Fixes #4327.

- [ ] this PR contains documentation
- [x] this PR contains tests
- [x] this PR contains YAML Samples

### CODE REVIEW

File: v2/specs/azure-rest-api-specs
Comment (theunrepentantgeek) [inline v2/specs/azure-rest-api-specs:1]: We should do the update of this submodule in a separate PR, will be easy to review and will reduce the size of this one.

File: v2/tools/generator/internal/codegen/pipeline/strip_descriptions.go [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+func createPropertyDescriptionRemovalVisitor() astmodel.TypeVisitor[any] {
+	visitor := astmodel.TypeVisitorBuilder[any]{
+		VisitObjectType: func(
+			this *astmodel.TypeVisitor[any],
+			it *astmodel.ObjectType,
+			ctx any,
+		) (astmodel.Type, error) {
+			result := it
+			for _, prop := range it.Properties().AsSlice() {
+				// If the property already has a description, use it
```
Comment (theunrepentantgeek) [inline v2/tools/generator/internal/codegen/pipeline/strip_descriptions.go:81]:
```suggestion
				// If the property already has an empty description, use it
```

File: v2/azure-arm.yaml [hunk 2]
```diff
@@ -1900,6 +1900,59 @@ objectModelConfiguration:
...
+          $payloadType: explicitEmptyCollections
+        Tags:
+          $payloadType: explicitEmptyCollections
+      ManagedClusterAgentPoolProfileProperties:
+        NodeLabels:
+          $payloadType: explicitEmptyCollections
+        NodeTaints:
+          $payloadType: explicitEmptyCollections
+        Tags:
+          $payloadType: explicitEmptyCollections
```
Comment (theunrepentantgeek) [inline v2/azure-arm.yaml:1937]: You can put `$payloadType` at any level, so you could put it directly on `ManagedClusterAgentPoolProfileProperties` and simplify, if all collections should be treated the same way.

File: v2/specs/azure-rest-api-specs
Comment (matthchr) [reply v2/specs/azure-rest-api-specs:1]: https://github.com/Azure/azure-service-operator/pull/4427

File: v2/tools/generator/internal/codegen/pipeline/strip_descriptions.go (hunk 1 above)
Comment (matthchr) [reply v2/tools/generator/internal/codegen/pipeline/strip_descriptions.go:81]: Fixed.

**Comment by matthchr** [comment]
I've changed how I stripped descriptions to make it a configuration option in `objectModelConfiguration` and done it per resoruce version, and only stripped some of the older versions documentation. Agree that's better - I had missed that it would impact our reference docs, which as you say we definitely don't want to lose.

**Pull request merged**